Add Pacman Package Cache docs (#101)
NUbots/NUbook: src/book/02-system/01-foundations/03-continuous-integration.mdx

![Screenshot of the build checks section of the PR page, with arrow pointing to primary link](./images/ci-build-logs.png 'Link to the build details with logs')

## Pacman Package Cache

When Buildkite receives a webhook to build and test a PR, the Buildkite agents pull down the code and start building a Docker image based off the Dockerfile in that branch.

In the process of building our Docker image, several [pacman](https://wiki.archlinux.org/index.php/pacman) packages are downloaded from public mirrors (if that specific [layer](https://docs.docker.com/storage/storagedriver/#images-and-layers) itself is not cached). Because these packages are downloaded numerous times, across multiple PRs, caching them locally can speed up build times and save network bandwidth.

> It is also possible that the [signature](https://wiki.archlinux.org/index.php/Pacman/Package_signing) on a package that is located on a public mirror changes. This change of signature can cause issues for us because we download version tagged packages and still expect an older signature. Caching such packages locally may alleviate these issues.

Reviewer: ```suggestion
It is also possible that the [signature](https://wiki.archlinux.org/index.php/Pacman/Package_signing) on a package that is located on a public mirror changes. This change of signature can cause issues for us because we download date-locked packages and still expect an older signature. Caching such packages locally may alleviate these issues.
```

We explicitly actually restrict the version of the packages, but instead we lock the date of the mirrors.

![Screenshot of the build checks section of the PR page, with arrow pointing to primary link](./images/ci-build-logs.png 'Link to the build details with logs')

## Pacman Package Cache

> When buildkite receives a webhook to build and test a PR, the buildkite agents pulls down the code and starts building a docker image based off the Dockerfile in that branch.

Reviewer 2: Capitalize Buildkite (like we do GitHub and others) and Docker (like they do in their own documentation). They should only be lowercase in things like commands.

```suggestion
When Buildkite receives a webhook to build and test a PR, the Buildkite agents pulls down the code and starts building a Docker image based off the Dockerfile in that branch.
```

There might be other instances on this page that need updating.

> In the process of building our docker image, serveral [pacman](https://wiki.archlinux.org/index.php/pacman) packages are downloaded from public mirrors (if that specific [layer](https://docs.docker.com/storage/storagedriver/#images-and-layers) itself is not cached). Because these packages are downloaded numerous times, across multiple PR's, caching them locally can speed up build times and save network bandwidth.

Reviewer: There have also been instances where the public mirrors have changed the signature on a package, but because we date lock our packages we are still expecting an older signature. Caching these package **_may_** alleviate this issue as well.

In order to cache these packages, a [pacman](https://wiki.archlinux.org/index.php/pacman) package cache is setup on each buildkite agent. This cache uses the [nroi/flexo](https://hub.docker.com/r/nroi/flexo) docker image, and exposes itself on

```
http://localhost:7878/$repo/os/$arch
```

> In order for [pacman](https://wiki.archlinux.org/index.php/pacman) to use this package cache, the cache must be added to `/etc/pacman.d/mirrorlist`. Because our build happens inside a [buildx](https://docs.docker.com/buildx/working-with-buildx/) container, we cannot use `localhost` as the address for the cache. Instead, because the buildx container runs on the `bridge` network, we can ping the host system on `172.17.0.1`.

Reviewer: Is this IP address guaranteed to always be this?

![Screenshot of the build checks section of the PR page, with arrow pointing to primary link](./images/ci-build-logs.png 'Link to the build details with logs')

## Pacman Package Cache

> When Buildkite receives a webhook to build and test a PR, the Buildkite agents pulls down the code and starts building a Docker image based off the Dockerfile in that branch.

Reviewer 2: ```suggestion
When Buildkite receives a webhook to build and test a PR, the Buildkite agents pull down the code and start building a Docker image based off the Dockerfile in that branch.
```

> In the process of building our docker image, serveral [pacman](https://wiki.archlinux.org/index.php/pacman) packages are downloaded from public mirrors (if that specific [layer](https://docs.docker.com/storage/storagedriver/#images-and-layers) itself is not cached). Because these packages are downloaded numerous times, across multiple PR's, caching them locally can speed up build times and save network bandwidth.

Reviewer 2: ```suggestion
In the process of building our Docker image, several [pacman](https://wiki.archlinux.org/index.php/pacman) packages are downloaded from public mirrors (if that specific [layer](https://docs.docker.com/storage/storagedriver/#images-and-layers) itself is not cached). Because these packages are downloaded numerous times, across multiple PRs, caching them locally can speed up build times and save network bandwidth.
```

> It is also possible that the [signature](https://wiki.archlinux.org/index.php/Pacman/Package_signing) on a package that is located on a public mirror changes. This change of signature can cause issues for us because we download version tagged packages and still expect an older signature. Caching these packages locally may alleviate these issues.

Reviewer 2: ```suggestion
It is also possible that the [signature](https://wiki.archlinux.org/index.php/Pacman/Package_signing) on a package that is located on a public mirror changes. This change of signature can cause issues for us because we download version tagged packages and still expect an older signature. Caching such packages locally may alleviate these issues.
```

> To cache these packages locally, a [pacman](https://wiki.archlinux.org/index.php/pacman) package cache is setup on each buildkite agent. This cache uses the [nroi/flexo](https://hub.docker.com/r/nroi/flexo) docker image, and exposes itself on

Reviewer 2: ```suggestion
To cache packages locally, a [pacman](https://wiki.archlinux.org/index.php/pacman) package cache is setup on each Buildkite agent. This cache uses the [nroi/flexo](https://hub.docker.com/r/nroi/flexo) Docker image, and exposes itself on:
```

```

> http://localhost:7878/$repo/os/$arch

Reviewer 2: Perhaps add a line below saying what `$repo` and `$arch` will be expanded to.

Author: Honestly not too sure what these expand to, they don't seem to be set as environment variables inside arch or our container, but they definitely do need to be in the URL.

Reviewer: Arch linux/pacman expands them internally. `$arch` is typically (and in our case will be) `x86_64`. `$repo` is probably one of these (and possibly others)
 - community
 - community-staging
 - community-testing
 - core
 - extra
 - multilib
 - multilib-staging
 - multilib-testing
 - staging
 - testing

Author: Since these are expanded internally by pacman and it isnt something we are concered with (plus couldnt find any documentation on them :stuck_out_tongue:), I dont think its necessary to add anything.

```

> In order for [pacman](https://wiki.archlinux.org/index.php/pacman) to use this package cache, the cache must be added to `/etc/pacman.d/mirrorlist`. Because our build happens inside a [buildx](https://docs.docker.com/buildx/working-with-buildx/) container, we cannot use `localhost` as the address for the cache. Instead, we alias the IP address of the host machine as `host.docker.internal`. Therefore, the following is added too the mirrorlist:

Reviewer 2: Use permalinks when linking to specific lines, as the lines might change in later commits. Press <kbd>y</kbd> when viewing a single file on GitHub to switch to the permalink.

```suggestion
In order for [pacman](https://wiki.archlinux.org/index.php/pacman) to use this package cache, the cache must be added to `/etc/pacman.d/mirrorlist`. Because our build happens inside a [buildx](https://docs.docker.com/buildx/working-with-buildx/) container, we cannot use `localhost` as the address for the cache. Instead, we alias the IP address of the host machine as `host.docker.internal`. Therefore, the following is added to the mirrorlist [here in the Dockerfile](https://github.com/NUbots/NUbots/blob/5149f1c0cd1e4ff902d749269d1aa90f52adc075/docker/Dockerfile#L14):
```

```
Server = http://host.docker.internal:7878/$repo/os/$arch
```

> 
> Adding mirror to the mirrorlist is done [here](https://github.com/NUbots/NUbots/blob/master/docker/Dockerfile#L14) in the Dockerfile.

Reviewer 2: ```suggestion
```

- <https://buildkite.com/docs/agent/v3/hooks>
- <https://buildkite.com/docs/agent/v3/configuration>

### Setup Pacman Package Cache

<Alert type="info">

> Whenever we update our base Arch Linux image [here in the Dockerfile](https://github.com/NUbots/NUbots/blob/5149f1c0cd1e4ff902d749269d1aa90f52adc075/docker/Dockerfile#L13), both [the mirror in the Dockerfile](https://github.com/NUbots/NUbots/blob/5149f1c0cd1e4ff902d749269d1aa90f52adc075/docker/Dockerfile#L14) **_and_** [the mirror in the cache config](https://github.com/NUbots/NUbots/blob/master/doc/PacmanCache/docker-compose.yml#L14) need to be updated.

Reviewer 3: we probably want to link to a specific commit instead of master for the third link too

- <https://buildkite.com/docs/agent/v3/hooks>
- <https://buildkite.com/docs/agent/v3/configuration>

### Setup Pacman Package Cache

> **NOTE:**
> Whenever we update our base archlinux image [here](https://github.com/NUbots/NUbots/blob/master/docker/Dockerfile#L13), both the mirror in the Dockerfile AND the mirror in the cache config need to be updated ([here](https://github.com/NUbots/NUbots/blob/master/docker/Dockerfile#L14) and [here](https://github.com/NUbots/NUbots/blob/master/doc/PacmanCache/docker-compose.yml#L14))

Reviewer: We have an `alert` tag for this purpose.
```suggestion
<Alert type="info">

Whenever we update our base archlinux image [here](https://github.com/NUbots/NUbots/blob/master/docker/Dockerfile#L13), both the mirror in the Dockerfile AND the mirror in the cache config need to be updated ([here](https://github.com/NUbots/NUbots/blob/master/docker/Dockerfile#L14) and [here](https://github.com/NUbots/NUbots/blob/master/doc/PacmanCache/docker-compose.yml#L14))

</Alert>
```

- <https://buildkite.com/docs/agent/v3/hooks>
- <https://buildkite.com/docs/agent/v3/configuration>

### Setup Pacman Package Cache

<Alert type="info">

> Whenever we update our base archlinux image [here](https://github.com/NUbots/NUbots/blob/master/docker/Dockerfile#L13), both the mirror in the Dockerfile AND the mirror in the cache config need to be updated ([here](https://github.com/NUbots/NUbots/blob/master/docker/Dockerfile#L14) and [here](https://github.com/NUbots/NUbots/blob/master/doc/PacmanCache/docker-compose.yml#L14))

Reviewer 2: Try to avoid links where the text is just "here".

```suggestion
Whenever we update our base Arch Linux image [here in the Dockerfile](https://github.com/NUbots/NUbots/blob/5149f1c0cd1e4ff902d749269d1aa90f52adc075/docker/Dockerfile#L13), both [the mirror in the Dockerfile](https://github.com/NUbots/NUbots/blob/5149f1c0cd1e4ff902d749269d1aa90f52adc075/docker/Dockerfile#L14) **_and_** [the mirror in the cache config](https://github.com/NUbots/NUbots/blob/master/doc/PacmanCache/docker-compose.yml#L14) need to be updated.
```

</Alert>

> 1. Create a directory to store cache config

Reviewer 2: ```suggestion
1. Create a directory to store the cache config
```

   ```
   mkdir ~/pacman-cache && cd ~/pacman-cache
   ```

> 2. Download install script

Reviewer 2: ```suggestion
2. Download the install script
```

   ```

>    wget https://raw.githubusercontent.com/NUbots/NUbots/master/doc/PacmanCache/install.sh

Reviewer 2: We should consider adding this script to [the agent setup script](https://github.com/NUbots/NUbook/blob/master/src/book/02-system/01-foundations/assets/ci_setup.sh).

Author: Probably a good idea, but also keep it as its separate script so people can set it up on their personal machines if they want.

   ```

> 3. Make install script executable

Reviewer 2: ```suggestion
3. Make the install script executable
```

   ```
   chmod +x ./install.sh
   ```

> 4. Run install script

Reviewer 2: ```suggestion
4. Run the install script
```

<Alert type="info">

> Whenever we update our base Arch Linux image [here in the Dockerfile](https://github.com/NUbots/NUbots/blob/5149f1c0cd1e4ff902d749269d1aa90f52adc075/docker/Dockerfile#L13), both [the mirror in the Dockerfile](https://github.com/NUbots/NUbots/blob/5149f1c0cd1e4ff902d749269d1aa90f52adc075/docker/Dockerfile#L14) **_and_** [the mirror in the cache config](https://github.com/NUbots/NUbots/blob/d38931c75c786647b211a597207f0ddf54dc5eef/doc/PacmanCache/docker-compose.yml#L14) need to be updated.

Reviewer: These links should be referencing pages in the `master` branch
